### DISCUSSION TITLE
Calibration dataset in INT8 Qunatization

### DISCUSSION THREAD

**Discussion by phanben110** [BlueMirrors/cvu#44] [Q&A] [answered]
Why does quantization INT8 need to have calibration dataset, but FP16, FP32 don't need ?

**Comment by mavihs7** [comment, marked as answer]
Hi @phanben110, 

Since we generally train models in FP32, we don't require any further calibration. And when you convert your model from FP32 to FP16, we just drop half of those 32bits and you loose some precision. 

However, if we do the same to go from FP16 to INT8, we'd loose all the precision and most of the numbers would just become zero. That's why instead of dropping bits, we map FP16 values to INT8 values to capture the range. And to capture this range properly, we run our FP16 models with a group of expected inputs values, which helps us in minimizing precision loss.  To know more about this topic, I'd recommend [this article](https://developer.nvidia.com/blog/achieving-fp32-accuracy-for-int8-inference-using-quantization-aware-training-with-tensorrt/)

Now what should be your calibration dataset, depends on your use case. Sometimes people just randomly sample from their training dataset, or sometimes they selectively sample for a specific use case. For example, let's say you trained a model with animal dataset where images were captured in all conditions (i.e. different weathers, low light, etc.). But you're deploying this model to certain parts of the forest and you expect to capture only certain types of images. So when doing the INT8 quantization, you'd want to calibrate with a more selective and relevant dataset to minimize precision loss for that specific use case.

Hope this helps.

**Reply by phanben110** [reply]
Thanks for your answer @mavihs7  ❤️, 

I have one more question to ask you. Have you ever done INT8 Quantization Aware Training task with yolov5s model? 

I have completed the Post Training Quantization task from your repository team, and I want to do the QAT (Quantization Aware Training) task for evaluation performance. 

Thanks!

**Reply by mavihs7** [reply]
Glad to be of help @phanben110 ! 

I haven't trained yolov5s in some time, so I'm not sure if ultralytics now support QAT. I would recommend you to directly ask this question on their official repo https://github.com/ultralytics/yolov5
